### PR TITLE
Multiple updates on workflow

### LINKED ISSUE #7133: GitHub Actions: Upgrade to Node 20 soon
Node 16 is no longer being supported and GitHub will deprecate it in Actions soon.

https://github.blog/changelog/2023-09-22-github-actions-transitioning-from-node-16-to-node-20

> Our plan is to transition all actions to run on Node 20 by Spring 2024. We will actively monitor the migration's progress and gather community feedback before finalizing the transition date.

We will want to transition to Node 20 or higher by updating workflows and actions across our repositories soon, or GitHub will transition for us.

@Hans5958, do you want to self-assign?

### LINKED ISSUE #7658: ESLint workflow not working
Error message on every run since... oh, I don't know when:
```
Run npx eslint . --config .eslintrc.json --ext .js
  npx eslint . --config .eslintrc.json --ext .js
  shell: /usr/bin/bash -e {0}
Invalid option '--ext' - perhaps you meant '-c'?
You're using eslint.config.js, some command line flags are no longer available. Please see https://eslint.org/docs/latest/use/command-line-interface for details.
Error: Process completed with exit code 2.
```
The check still succeeds, it just can't evaluate anything. (That's probably why we haven't noticed.)

It might also be a good time to revise the rules, see if we should be more restrictive about certain things like unused variables.

@WorldLanguages @Hans5958

### PR DESCRIPTION
**Pull request by Hans5958**
Resolves #7133 
Resolves #7658  
Relates to #7116

### Changes

Some updates on the workflows, especially on bumping the dependencies. All bumps are considered by looking at each and every changelog.

### Reason for changes

Housekeeping

### Tests

It should've worked, but there may some tests that would need to be done, especially on our in-house scripts.

### CODE REVIEW

**Comment by Hans5958** [comment]
@apple502j Would it be possible to run unittest for every push/PR, or are there any considerations?

I was thinking we can turn off the "create issue when failed," and just give a failure when it fails.

**Review by DNin01 — COMMENTED** [review]
I'm not super familiar with the language of workflows or actions so I can't tell if these upgrades will break anything, but everything else looks good.

- Nice dependency updates.
- Updating Prettier shouldn't be a huge deal for contributors because we added `prettier_version`. For us, just remember to re-run Prettier on the source repo so it's all green.
- Nice new, concise name for the Transifex workflow.

**Comment by Hans5958** [comment]
I had an idea to update the l10n workflow so it beautifies the JSON files like what I did on [website-i18n](https://github.com/ScratchAddons/website-i18n/commit/c0674f4e4701da4da99859a7da79798b633282e7), instead of it being minifed, but then I would increase the extension package quite a bit.

> Note that apple502j is "on inedible hiatus with regard to SA for a while", so he might not be able to respond.

Hmm, fair. Should I just wing it and do that? I saw that it only checks l10n files, maybe we should expand it more...

**Comment by DNin01** [comment]
> I had an idea to update the l10n workflow so it beautifies the JSON files like what I did on [website-i18n](https://github.com/ScratchAddons/website-i18n/commit/c0674f4e4701da4da99859a7da79798b633282e7), instead of it being minifed, but then I would increase the extension package quite a bit.

Well, would this reduce the file size of the diffs? It would make them more readable for sure.

I did this with one of the repositories I store locally, but I also trimmed all the indentation which reduced the file size quite a bit after formatting. Of course, the file itself is not as small as it would be all in one line, and the trick might be more effective on JSON with lots of nested objects, but it's a good compromise for having small, readable diffs.

What I mean is something like this:
```json
{
"a": {
"b": "c"
}
}
```

File: .github/workflows/transifex.yml [hunk 1]
```diff
@@ -1,20 +1,21 @@
-name: Pull and push strings via Transifex
+name: Sync strings via Transifex
```
Comment (DNin01) [inline .github/workflows/transifex.yml:1]: Nice change. But usually there's something to synchronize data _with_.
```suggestion
name: Sync strings with Transifex
```
I don't know if that really matters.

**Comment by WorldLanguages** [comment]
> I had an idea to update the l10n workflow so it beautifies the JSON files

I would try to avoid changes like this. They're not really necessary.

**Comment by Samq64** [comment]
> > I had an idea to update the l10n workflow so it beautifies the JSON files
> 
> I would try to avoid changes like this. They're not really necessary.

Being minified also discourages newer contributors from modifying them directly.

File: .github/workflows/transifex.yml (hunk 1 above)
Comment (Hans5958) [reply .github/workflows/transifex.yml:1]: Yeah, I forgot that you said to use "with" after I pushed the change, but I'm still being undecided.

Now you mentioned it, I would change it to it.

**Comment by Hans5958** [comment]
> I would try to avoid changes like this. They're not really necessary.

If your concern is it being too minor, it's more of because "since we're doing a bundle of updates, might as well include more updates."

File: .github/workflows/format-code.yml
Comment (DNin01) [inline .github/workflows/format-code.yml:1]: Is it possible to only run Prettier formatting when certain paths are updated? We already ignore generated files like locales other than English, so running formatting when those are changed won't do anything.

I don't know if there's a way to make a _blacklist_ of paths.

Comment (Hans5958) [reply .github/workflows/format-code.yml:1]: Yes, you can, with `paths-ignore`. I have did it, and also a message to sync both paths.

**Comment by Hans5958** [comment]
By the way, this PR is _really_ ready for review. I hope this can be merged in a not so long time.

**Comment by DNin01** [comment]
Can `actions/upload-artifact` also be updated to v4?

**Comment by Hans5958** [comment]
I mean, this MR is way, way, **way** overdue.

**Comment by Hans5958** [comment]
Also updated the ESLint config to be able to be used on ESLint v9.x. Now this PR resolves #7658.

**Review by DNin01 — COMMENTED** [review]
I'm not well versed in this sort of stuff, but this looks good AFAIK.

**Comment by DNin01** [comment]
We have until June 30th, unless we opt into continuing to use Node 16.

> On June 30th 2024, we will change the default from Node16 to Node20.
> To opt out of this and continue using Node16 while it is still available in the runner, you can choose to set `ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION=true`
> 
> We will then continue to monitor Node16 usage and will communicate a timeline for the removal of Node16 at the start of October, based on the volume of continued use. This means that customers who use the environment variable to continue to use Node16 now have until October to complete their migrations.

https://github.blog/changelog/2024-05-17-updated-dates-for-actions-runner-using-node20-instead-of-node16-by-default/

File: .github/workflows/eslint.yml [hunk 2]
```diff
@@ -1,28 +1,31 @@
...
-      - name: Install ESLint
-        run: |
-          npm install eslint
+        uses: actions/checkout@v4
 
       - name: Run ESLint
-        run: npx eslint .
-          --config .eslintrc.json
-          --ext .js
+        run: npx eslint . --config .eslintrc.json --ext .js
```
Comment (DNin01) [inline .github/workflows/eslint.yml:30]: Was this supposed to be updated to refer to `eslint.config.mjs`?

Comment (DNin01) [reply .github/workflows/eslint.yml:30]: Also, I don't know if we even have to specify `--config`.

**Comment by Hans5958** [comment]
I forgot to test if the ESLint is working. Turns out I still have adjustments to do. I also have to add the `npm install` bodge, unless I can do `package.json`. This should work for now.

@WorldLanguages, should I add `package.json`?

> We have until June 30th, unless we opt into continuing to use Node 16.

Am I missing something? It's been a while and the workflows have been updated to v20.x. For the note, since June 30 is the past, the workflows that uses it will use Node 20 regardless.

File: .github/workflows/eslint.yml (hunk 2 above)
Comment (Hans5958) [reply .github/workflows/eslint.yml]: `--config` is not really needed, but I just left it *just in case*.

**Comment by DNin01** [comment]
> > We have until June 30th, unless we opt into continuing to use Node 16.
> 
> 
> 
> Am I missing something? It's been a while and the workflows have been updated to v20.x. For the note, since June 30 is the past, the workflows that uses it will use Node 20 regardless.

Oh, it's not June anymore 😄

File: eslint.config.mjs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        Chart: "readonly",
+        Comlink: "readonly",
+        idb: "readonly",
+        Scratch: "readonly",
+        Blockly: "readonly",
+        Clipboard: "readonly",
+        ClipboardItem: "readonly",
+      },
+
+      ecmaVersion: 2020,
```
Comment (Samq64) [inline eslint.config.mjs:43]: The oldest browser we now support is Chrome 96 from November 2021.
```suggestion
      ecmaVersion: 2021,
```
The comment near the end of the file should also be updated.

Comment (Hans5958) [reply eslint.config.mjs:43]: Is there the information for the minimum version of Firefox, just so I can do it in one go?

Comment (Samq64) [reply eslint.config.mjs:43]: It's much more recent, 109 from January 2023.

**Comment by DNin01** [comment]
Deadline to update `actions/upload-artifact`: November 30th.

**Review by DNin01 — APPROVED** [review]
I'm not familiar enough with CI to do a thorough review, but I couldn't find any glaring issues, and this is long overdue, so I think we're good if you agree.

**Comment by Samq64** [comment]
Could ESLint be moved to a separate PR? It does not work without `globals` installed and the easiest proper fix is to introduce a `package.json` which will need some more discussion.

**Comment by DNin01** [comment]
The date has changed:
> Artifact actions v3 will be closing down by **December 5, 2024**. To raise awareness of the upcoming removal, we will temporarily fail jobs using v3 of `actions/upload-artifact` or `actions/download-artifact`. Builds that are scheduled to run during the brownout periods will fail. The brownouts are scheduled for the following dates and times:
> – November 14, 12pm – 1pm EST
> – November 21, 9am – 5pm EST

https://github.blog/changelog/2024-11-05-notice-of-breaking-changes-for-github-actions/#artifacts-v3-brownouts

**Comment by Hans5958** [comment]
The ESLint fix is for #7658. Also, I don't expect for people to use ESLint out of the box (esp. when we don't define the version to be used and the latest ESLint won't work AFAIK), so that's why I upgrade it.

**Comment by DNin01** [comment]
@Hans5958 We don't have much longer. Should we just merge?

**Comment by Hans5958** [comment]
I hope I have said this, but `@actions/upload-artifact` is only used on a rarely used workflow action, so even if we lapse it, the risk is low.

**Comment by DNin01** [comment]
> I hope I have said this, but `@actions/upload-artifact` is only used on a rarely used workflow action, so even if we lapse it, the risk is low.

We're getting ready to release v1.40.0, though...?

**Comment by DNin01** [comment]
It'd be nice if we had Dependabot here.

**Comment by DNin01** [comment]
@WorldLanguages, we're cutting it kinda close here if you're planning to release v1.40.0 soon. If we don't upgrade `upload-artifact` by merging this PR before December 5th, it'll stop working and we won't be able to release any more updates after that date.

**Comment by Hans5958** [comment]
> > I hope I have said this, but `@actions/upload-artifact` is only used on a rarely used workflow action, so even if we lapse it, the risk is low.
> 
> We're getting ready to release v1.40.0, though...?

CMIIW, but I thought that is done manually, not here?

**Comment by DNin01** [comment]
> > > I hope I have said this, but `@actions/upload-artifact` is only used on a rarely used workflow action, so even if we lapse it, the risk is low.
> > 
> > 
> > We're getting ready to release v1.40.0, though...?
> 
> CMIIW, but I thought that is done on WL's PC, not here?

I saw [this action](https://github.com/ScratchAddons/ScratchAddons/actions/workflows/packer.yml) which WL runs every release to generate builds to attach to the [releases](https://github.com/ScratchAddons/ScratchAddons/releases).

**Comment by Hans5958** [comment]
Huh, didn't saw it is last run on v1.39.x.

Okay, I would use my discretion and will merge this in two days in the earliest, assuming no significant changes are being requested.

**Review by Samq64 — APPROVED** [review]
Untested

**Comment by DNin01** [comment]
> Can't seem to do this since I don't have a power to bypass it, though.

You don't have permission to bypass it? Is that because you aren't on the @ScratchAddons/admins team? But you're an organization owner, so…?

But Samq just approved, so it should be able to be merged now.

**Comment by DNin01** [comment]
> > If I can, I would use my discretion and will merge this by myself. Can't seem to do this since I don't have a power to bypass it, though.
>
> You don't have permission to bypass it? Is that because you aren't on the @ScratchAddons/admins team? But you're an organization owner, so…?

Psst... this might be something you could fix, @WorldLanguages.

<details><summary>Possible solutions</summary>
<p>

1. Check that the "Do not allow bypassing the above settings" option in the branch protection isn't checked.
2. Add Hans to the team.
3. Create a ruleset — which are like branch protections 2.0 — that can be bypassed by organization owners (let me know if you need help with this, I've used rulesets before).

</p>
</details>

**Comment by Hans5958** [comment]
I will merge it after checking the files again on my PC. I'm typing this on my phone.

IIRC The admin team is just a team with additional perms to ping every admin, that is people who has Owner access to the organization. I'm just left the group to avoid any notifications since I'm pretty inactive at that time, but I still have the Owner role, and somehow I don't have access on bypass the check. It's weird. I may check it sometime later, I guess.

**Comment by Hans5958** [comment]
Somehow Prettier workflow fails on master? I will take a look.

**Comment by DNin01** [comment]
ESLint seems to have detected some errors also. They have probably been accumulating from when the scanning wasn't working.

Even then, the check was still a success, somehow. Might need to do some configuration.

**Comment by WorldLanguages** [comment]
@DNin01 We figured out why Hans couldn't merge this as admin - he had the perms to add himself to the 'allowed list', but he wasn't actually in the list.

By the way, I added @mxmou and @Samq64 to that list, so that they can merge pull requests where they are the PR authors and there is 1 additional approval, which we already consider to equal 2 org approvals, even if GitHub doesn't agree.

**Comment by WorldLanguages** [comment]
Unfortunately, the 'Pack extensions' action is still not working as expected.

`No files were found with the provided path: .dist/*.zip. No artifacts will be uploaded.`

I will pack the extension manually this time

**Comment by Samq64** [comment]
> `No files were found with the provided path: .dist/*.zip. No artifacts will be uploaded.`

Most likely this is caused by https://github.com/actions/upload-artifact/issues/602.

**Comment by DNin01** [comment]
> I will pack the extension manually this time

Phew. Why not always do it manually anyway?

**Pull request merged**